common/djangoapps/entitlements/models.py

```diff
@@ -320,6 +329,99 @@ def get_active_entitlements_for_user(cls, user):
             enrollment_course_run=None
         ).select_related('user').select_related('enrollment_course_run')
 
+    @classmethod
+    def get_fulfillable_entitlements(cls, user):
+        """
+        Returns all fulfillable entitlements for a User
+
+        Arguments:
+            user (User): The user we are looking at the entitlements of.
+
+        Returns
+            list<CourseEntitlements>: A list of course Entitlements that a user can enroll in. These must not be expired
+                and not have a course run already assigned to it.
+        """
+
+        try:
+            return cls.objects.filter(
+                user=user,
+            ).exclude(
+                expired_at__isnull=False,
+                enrollment_course_run__isnull=False
+            ).order_by('created')
+        except CourseEntitlement.DoesNotExist:
+            return []
+
+    @classmethod
+    def get_fulfillable_entitlement_for_user_course_run(cls, user, course_run_key):
+        """
+        Retrieves a fulfillable entitlement for the user and the given course run.
+
+        Arguments:
+            user (User): The user that we are inspecting the entitlements for.
+            course_run_key (CourseKey): The course run Key.
+
+        Returns:
+            CourseEntitlement: The most recent fullfillale CourseEntitlement, None otherwise.
+        """
+        # Check if the User has any fulfillable entitlements.
+        # Note: Wait to retrieve the Course UUID until we have confirmed the User has fulfullable entitlements.
+        # This was done to avoid calling the APIs when the User does not have an entitlement.
+        entitlements = cls.get_fulfillable_entitlements(user)
+        if entitlements:
+            course_uuid = get_course_uuid_for_course(course_run_key)
+            if course_uuid:
+                for entitlement in entitlements:
```

Comment:
rather than looping through all of the user's entitlements here you can filter the queryset returned from `get_fulfillable_entitlements` by course_uuid.